tutorials/lesson3-anatomy/coin.template.qnt

```diff
@@ -0,0 +1,953 @@
+    !*/
+
+    // The temporal property that says the following:
+    // If `NoSupplyOverflow` holds in a state, then this state may not
+    // produce a computation that ends in another state,
+    // where totalSupplyDoesNotOverflowInv is violated.
+    //
+    // Temporarily disabled, until the issue below is fixed:
+    // [#591](https://github.com/informalsystems/quint/issues/591).
+    //
+    // temporal NoSupplyOverflow: bool = always(totalSupplyDoesNotOverflowInv)
+    /*!
+      <step>
+        <title>Temporal properties</title>
+        <description>
+
+After we have defined state invariants, we should normally think about
+temporal properties in general such as safety and liveness. However, 
+temporal properties are an advanced topic, and smart
+contracts are typically focused on safety. Thus, it is safe to skip this topic
+for now. It is just important to know that temporal properties are labelled
+with the qualifier `temporal`.
+
+        </description>
+        <file>$file</file>
+        <line>$line</line>
+      </step>
+    !*/
+
+    // TESTS
+
+    // send should not work without minting
+    run sendWithoutMintTest = {
+        init.then(send(minter, "bob", 5))
+            .fail()
+    }
+    /*!
+      <step>
+        <title>Tests</title>
+        <description>
+
+So far, we have been running sequences of actions in REPL, to get
+basic understanding of the protocol mechanics. While REPL is capable of
+replaying actions one-by-one, it would be more convenient to run something
+similar to unit tests or integration tests, which are ubiquitous in
+programming languages.
+
+We normally add tests in the very bottom of the protocol module, or in
+a separate module.
+
+Quint introduces runs to express "happy paths" and tests. The above code
+shows a simple test `sendWithoutMintTest`.
+In this test, the action `init` runs first. If it evaluates to `true`,
+then the action `send(...)` is run. Since this action is composed with
+`fail()`, the action `send(...)` is expected to evaluate to `false`.
+
+Go ahead and see, whether this test goes through:
+
+            <run target="markdown">echo 'sendWithoutMintTest' | quint -r coin.qnt::Coin</run>
+            <run target="codetour">echo 'sendWithoutMintTest' | quint -r ./lesson3-anatomy/coin.qnt::Coin</run>
+
+            <block>
+**Exercise:** Actually, if you look carefully at the code of `send`,
+you can find one value for `amount` that makes `send` work even without
+prior minting. What is this value?
+            </block>
+        </description>
+        <file>$file</file>
+        <line>$line</line>
+      </step>
+    !*/
+
+    // `mint`, then `send`
+    run mintSendTest = {
+        init.then(mint(minter, "bob", 10))
+            .then(send("bob", "eve", 4))
+            .then(all {
+                assert(balances.get("bob") == 6),
+                assert(balances.get("eve") == 4),
+                minter' = minter,
+                balances' = balances,
+            })
```

Comment:
This is basically a homebrewed postcondition. There should probably be a keyword, like `fail()`, e.g. `expect(cond)` that does this.:

```
    run mintSendTest = {
        init.then(mint(minter, "bob", 10))
            .then(send("bob", "eve", 4))
            .expect( all {
                balances.get("bob") == 6,
                balances.get("eve") == 4
            })
    }
```

Reply:
yeah, I am trying to settle on a good set of operators for tests. So far we have `assert(...)`, `fail()`, and potentially, `ok()` #594. I like your idea of `expect` quite a lot, as it would let us to converge on a single useful operator.